### ISSUE TITLE
SocketsHttpHandler.MaxResponseHeadersLength can cause internal 32-bit overflow with HTTP/2

### ISSUE THREAD

**Issue by JamesNK** [dotnet/runtime#73848]
### Description

Noticed here: https://github.com/grpc/grpc-dotnet/issues/1842#issuecomment-1212059692

MaxResponseHeadersLength is in kilobytes so [internally multiplies the value by 1024](https://github.com/dotnet/runtime/blob/08a11e4a8383605087429df2477dc6a79e3ed8be/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L106).

However, the value it is stored on for HTTP/2 is 32 bit: https://github.com/dotnet/runtime/blob/08a11e4a8383605087429df2477dc6a79e3ed8be/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L88

HTTP/3 is using 64 bit:
https://github.com/dotnet/runtime/blob/80640d4d28b72ddcb3e8a69e49d88f644484b25f/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs#L41

### Reproduction Steps

```cs
new SocketsHttpHandler()
  {
    MaxResponseHeadersLength = 20 * 1024 * 1024
  };
```

Then do some HTTP/2 requests.

### Expected behavior

Either works, or provides a friendly argument exception when setting MaxResponseHeadersLength that the value exceeds the maximum size.

### Actual behavior

Internal overflow of 32-bit value, unpredictable behavior, bad error message.

```
      System.Net.Http.HttpRequestException: An error occurred while sending the request.
       ---> System.IO.IOException: The request was aborted.
       ---> System.Net.Http.HttpRequestException: The HTTP response headers length exceeded the set limit of 21474836480 bytes.
         at System.Net.Http.Http2Connection.Http2Stream.OnHeader(ReadOnlySpan`1 name, ReadOnlySpan`1 value)
         at System.Net.Http.Http2Connection.Http2Stream.System.Net.Http.IHttpHeadersHandler.OnStaticIndexedHeader(Int32 index)
         at System.Net.Http.HPack.HPackDecoder.OnIndexedHeaderField(Int32 index, IHttpHeadersHandler handler)
         at System.Net.Http.HPack.HPackDecoder.Parse(ReadOnlySpan`1 data, Int32& currentIndex, IHttpHeadersHandler handler)
         at System.Net.Http.HPack.HPackDecoder.DecodeInternal(ReadOnlySpan`1 data, IHttpHeadersHandler handler)
         at System.Net.Http.Http2Connection.ProcessHeadersFrame(FrameHeader frameHeader)
         at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
         --- End of inner exception stack trace ---
         at System.Net.Http.Http2Connection.ThrowRequestAborted(Exception innerException)
         at System.Net.Http.Http2Connection.Http2Stream.CheckResponseBodyState()
         at System.Net.Http.Http2Connection.Http2Stream.TryEnsureHeaders()
         at System.Net.Http.Http2Connection.Http2Stream.ReadResponseHeadersAsync(CancellationToken cancellationToken)
         at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
         at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
         at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
         at Grpc.Shared.TelemetryHeaderHandler.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
         at Grpc.Net.Client.Balancer.Internal.BalancerHttpHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         at Grpc.Net.Client.Internal.GrpcCall`2.RunCall(HttpRequestMessage request, Nullable`1 timeout)
```

**Comment by MihaZupan** [comment]
Technically a bug, but if you're setting this so high you're pretty much begging to be DOSed.

**Comment by AnashOommen** [comment]
@MihaZupan could you kindly explain why a high value is begging to be DOSed? In Google Ads API (which uses grpc), we return serialized server errors to the caller as trailing metadata, and that data can be more than a few KB. How can a user DOS themselves in this scenario?

**Comment by MihaZupan** [comment]
A few KB is perfectly fine.
The issue is when you are telling the client that it's okay to receive MBs or GBs of headers on a single response.
All of this data has to be buffered in memory (`HttpResponseMessage.TrailingHeaders` for example).

A malicious server could feed you GBs of data that the client would be forced to keep in memory all at once.

**Comment by AnashOommen** [comment]
@MihaZupan gotcha. Can we fix the underlying overflow by throwing an exception when trying to set a large value? The way this fails now isn't very helpful.

**Comment by MihaZupan** [comment]
I don't thing we'd want to start throwing now given that we've been [silently handling it for HTTP/1.1](https://github.com/dotnet/runtime/blob/21dc0040b5d6cd52501d9c502fb3ddbc23db97d7/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L558) so far.

We would likely fix HTTP/2 to behave the same way as 1.1 (cap at `int.MaxValue`), and potentially consider an analyzer to warn you that the value being set doesn't make any sense.

**Comment by AnashOommen** [comment]
Yeah, capping + analyzer works fine as well.
